refactor(ui-concerto): refactor concerto form to use React hooks

DianaLease opened the pull request:
### Changes
- refactor(ui-concerto): refactor concerto form to use React hooks

### Flags
<!--- Provide context or concerns a reviewer should be aware of -->
- @dselman - To note: I kept some of the keys with the html element type prefixed, since otherwise I received an error `Encountered two children with the same key. Keys should be unique so that components maintain their identity across updates. Non-unique keys may cause children to be duplicated and/or omitted — the behavior is unsupported and could change in a future version` even though they were of different html element types.
- **BREAKING CHANGE** - the passed in `onValueChange` prop should NOT be used to update the `json` prop passed to ConcertoForm! ConcertoForm updates its own value state internally and should only use the value passed as `json` for its initial render.

### Author Checklist
- [ ] Vital features and changes captured in unit and/or integration tests
- [x] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [ ] Extend the documentation, if necessary
- [ ] Merging to `master` from `fork:branchname`
- [x] Manual accessibility test performed
    - [x] Keyboard-only access, including forms
    - [ ] Contrast at least WCAG Level A
    - [ ] Appropriate labels, alt text, and instructions

dselman commented:
I think that the tests needs to be updated. If you run `npm test` inside the `ui-concerto` package and then select the `all` option you will see some test failures. We should also figure out why the tests are not running as part of the build.

dselman commented on packages/ui-concerto/src/lib/components/concertoForm.js:
```diff
@@ -13,213 +13,156 @@
-  removeElement(e, key, index) {
-    const array = get(this.state.value, key);
+const ConcertoForm = (props) => {
+  const [value, setValue] = useState(typeof props.json === 'string' ? JSON.parse(props.json) : JSON.parse(JSON.stringify(props.json)));
```
It seems quite odd that this `json` property accepts both JSON and a String. I wonder if there is a way to make this cleaner with the Storybook so we can remove this?

DianaLease replied:
If this was only a limitation of the storybook, we can only allow true objects here, since the storybook does not dictate the json anymore anyway.

dselman commented on packages/ui-concerto/src/lib/components/concertoForm.js:
```diff
@@ -13,213 +13,156 @@
+      const decodedModel = decodeHTMLEntities(model);
+      modelManager.addModelFile(decodedModel, `model-${idx}`, true);
+    });
+    modelManager.updateExternalModels().then(() => {
```
I think this we used to have an option to disable the download of external models?

dselman commented on packages/ui-concerto/src/lib/components/concertoForm.js:
```diff
@@ -13,213 +13,156 @@
-        return (
-          <Form style={{ minHeight: '100px', ...this.props.style }}>
-            {this.generator.generateHTML(this.props.type, this.state.value)}
+  if (props.type && value) {
```
Can `props.type` ever be null?

DianaLease replied:
I added a new warning message if there is no `props.type`.

dselman commented on packages/ui-concerto/src/lib/components/concertoForm.js:
```diff
@@ -13,213 +13,156 @@
 
 ConcertoForm.propTypes = {
   models: PropTypes.arrayOf(PropTypes.string),
   type: PropTypes.string,
```
Most of these are now required?

dselman commented on packages/ui-concerto/src/lib/reactformvisitor.js:
```diff
@@ -46,6 +44,8 @@ import {
+const toPath = (paramsArr) => paramsArr.join('.');
+
 /**
  * Returns true if the field has the @FormEditor( "hide", true )
```
Did you re-test with hidden fields, to ensure we don't get `null` children?

DianaLease replied:
We will have a `null` child if a field is hidden, but that's the correct approach: https://reactjs.org/docs/conditional-rendering.html#preventing-component-from-rendering.

DianaLease commented:
I addressed the feedback, but I'm still having trouble with the tests. @dselman or @mttrbrts, if one of you has time to pair on those tomorrow, I would be very grateful!

dselman reviewed (approved):
LGTM!
